RFC: Changes for fedora/anaconda systemd-boot functionality. Consolidate EFI/systemd directory naming.

jlinton opened the pull request:
I got stepped on with the 'efi-boot-options-description' which duplicates some functionality I added, as well as a couple other minor tweaks. As such here are the changes i'm using at the moment to support anaconda installing systemd-boot as the default bootloader on aarch64 machines.

jlinton commented:
I've also been considering making the EFI/systemd default a compile time constant.

poettering commented:
So the first commit should become redundant with #24777. 

The second commit sounds wrong to me, the idea really is that there's a *single* boot loader per ESP installed, but it can boot into boot menu items supplied by many different distros. Can you explain the rationale behind your change?

jlinton commented:
The first one is also potentially made redundant by Ard's zboot changes landing because the distro can choose to enable that rather than continue with the current choice of just having compressed efi binaries. But that doesn't mean its not helpful for users/distros that think that is the best plan (having a distro that is still grub focused and is compressing the efi image).

The second one gives people a choice, as today, where there are multiple grubs, it's possible that one or another distro/etc will have a signed image but not others. I think the idea that there is a single unified binary to boot multiple distro's sounds like a good idea, i'm just not sure that is where we are today.

keszybz reviewed (changes requested):
It'd help a lot if you split out one of the patches into a separate pull request. It'll make it easier to discuss things.

I made some comments on the implementation as I was reading the diff, in case we decide to use some parts of this. But let's have the high-level discussion first before spending time on the implementation.

About the first commit:
I think we might want to carry it as a downstream patch, while we're waiting for other things to get merged in other places. So it'd be nice to get it into proper form even if we don't end up merging it here.

About the second commit:
I don't grok the motivation. It makes (some) sense to have multiple grubs, because each distro has a different grub patchset, so there *are* different grubs. But there's just one systemd-boot, and all distro have the same code. The only thing that could be different between distros would be signing: e.g. Ubuntu has a signature that shim likes but Fedora does not. But even in that case, it's fine to have just one systemd-boot copy, because you can use the signed copy wherever. And considering the case where different distros would have systemd-boot with different signatures: it's enough to use one or the other to boot multiple OSes. systemd-boot itself does not embed any signatures, that's all external, so any signed copy of systemd-boot should work.

jlinton commented on src/boot/bootctl.c:
```diff
@@ -2293,7 +2310,7 @@ static int verb_remove(int argc, char *argv[], void *userdata) {
         }
 
-        char *path = strjoina("/EFI/systemd/systemd-boot", get_efi_arch(), ".efi");
+        char *path = strjoina(arg_loader_path,"systemd-boot", get_efi_arch(), ".efi");
```
First, thanks for looking at this... And your right it needs some discussion first, which is why I didn't remove the bashisms yet. It's possible the zboot stuff will land in fedora and be enabled soon, so that part may not even need to be downstream.

So, I can see the want/need to have a single systemd-boot, but I just don't see how it's going to work. Right now few machines have distro enrolled keys, and I can't guess which vendors might enroll which keys on which hardware, and even if MS signs a db update with some portion of the distro keys I suspect there will remain a situation where only a subset of the distro's keys are in place on any given machine. And while systemd-boot may remain fairly consistent between distros' unless a 3rd key just for the bootloader is in the firmware the risk exists if say systemd-boot is signed by fedora, and is being used to boot fedora as well as another distro (signed with a different key) and the fedora key needs to be revoked/whatever that the other distro would expect an alternate boot path to remain. Never mind the situation about which distro gets to own the systemd-boot used to boot them, or the naming. Given the other patch to update the UEFI name means that the UEFI boot selection menus themselves are likely going to read $DISTRO.  It seems a big mess of issues just to avoid having a couple of copies of systemd-boot in different EFI/$DISTRO directories. Multiple copies aren't great, but in my mind it simplifies the boot process, each distro gets its own efi entry, and its own key used to validate the entire boot path. The loaders/entries allow each systemd-boot to possibly boot other distro's from a single menu if the keys are setup right.

So, as I mentioned in the second comment, I don't like the way this set behaves. It works for installing/etc, but I would prefer that if its installed to an alternate location that the tool is a bit sticker about it. So I'm probably going to rework it to have a couple compile time defaults/etc. Or maybe if everyone thinks its a bad idea, then we can try sharing a single loader...

bluca replied:
I'm not sure I follow the use case. As explained earlier, sd-boot does not embed any key/signature/hash. So if you don't want to take shim+sd-boot from the same vendor (so that the key embedded in shim verifies sd-boot), you can simply ensure that whatever key is used to sign sd-boot is enrolled in MoK. Then which shim is used no longer matters (well, there's the SBAT stuff coming, but that's orthogonal).

jlinton commented:
I've split this into two branches. This branch is just the EFI/systemd vs EFI/distroname change. I've removed the command line parameter instead, largely focusing on consolidating the EFI/systemd into a single string, which makes downstream patching easier should a distro choose to utilize its own name. The second patch needed to deal with current distro's compressed kernel images I've moved to another branch.

jlinton commented:
The arm specific code is now here: https://github.com/systemd/systemd/pull/25309

poettering commented:
So yeah, as @keszybz and @bluca I don't grok the rationale for moving this to distro-specific dirs given this is supposed to be distro-agnostic: a single sd-boot should be able to load any distro.

Yes, SecureBoot makes such multi-boot scenarios a bit harder, but mokutil exists, and that's how it should be solved: shim will come with one set of keys built-in but you can add in others dynamically to allow multi-boot with SecureBoot.
